examples/experimental/dagster-airlift/dagster_airlift/in_airflow/partition_utils.py

```diff
@@ -0,0 +1,124 @@
+            raise Exception(
+                "Found some unpartitioned assets and some partitioned assets in the same task. "
+                "For a given task, all assets must have the same partitions definition. "
+            )
+        partition_keys_per_asset = [
+            set(asset_node["partitionKeys"])
+            for asset_node in asset_nodes
+            if asset_node["isPartitioned"]
+        ]
+        if not all_sets_equal(partition_keys_per_asset):
+            raise Exception(
+                "Found differing partition keys across assets in this task. "
+                "For a given task, all assets must have the same partitions definition. "
+            )
+        # Now we can proceed with the assumption that all assets are partitioned and have the same partition keys.
+        # This, we only look at the first asset node.
+        asset_node = next(iter(asset_nodes))
+        if not asset_node["isPartitioned"]:
+            return None
+        partitioning_type = PartitionDefinitionType(asset_node["partitionDefinition"]["type"])
+        return PartitioningInformation(
+            partitioning_type=partitioning_type,
+            partition_keys=asset_node["partitionKeys"],
+            additional_info=_build_additional_info_for_type(asset_node, partitioning_type),
+        )
+
+    @property
+    def time_window_partitioning_info(self) -> TimeWindowPartitioningInformation:
+        if self.partitioning_type != PartitionDefinitionType.TIME_WINDOW:
+            raise Exception(
+                f"Partitioning type is {self.partitioning_type}, but expected {PartitionDefinitionType.TIME_WINDOW}"
+            )
+        if self.additional_info is None:
+            raise Exception(
+                f"Partitioning type is {self.partitioning_type}, but no additional info was provided."
+            )
+        return self.additional_info
+
+
+def _build_additional_info_for_type(
+    asset_node: Mapping[str, Any], partitioning_type: PartitionDefinitionType
+) -> Optional[TimeWindowPartitioningInformation]:
+    if partitioning_type != PartitionDefinitionType.TIME_WINDOW:
+        return None
+    return TimeWindowPartitioningInformation(fmt=asset_node["partitionDefinition"]["fmt"])
+
+
+def all_sets_equal(list_of_sets):
+    if not list_of_sets:
+        return True
+    return len(set.union(*list_of_sets)) == len(set.intersection(*list_of_sets))
+
+
+def translate_logical_date_to_partition_key(
+    logical_date: datetime, partitioning_info: PartitioningInformation
+) -> str:
+    if not partitioning_info.partitioning_type == PartitionDefinitionType.TIME_WINDOW:
+        raise Exception(
+            "Only time-window partitioned assets or non-partitioned assets are supported out of the box."
+        )
+    fmt = partitioning_info.time_window_partitioning_info.fmt
+    partitions_and_datetimes = [
+        (_get_partition_datetime(partition_key, fmt), partition_key)
```

Comment:
Just to write out one potential downside as there should be potential version skew between the implementation of datetime in the dagster process and in the airflow process (I suspect it has been quite stable over time).

And just in general passing a format string that can drive compute can be dangerous via a format string attack. There have been vulnerabilities in C. (e.g. https://owasp.org/www-community/attacks/Format_string_attack).

Reply:
However in this case although the string comes from an external we can probably consider the dagster process a non-external "source"

Reply:
my thinking as well. I guess if someone somehow hijacks the endpoint it can be problematic.